include/uk/arch/spinlock.h

```diff
@@ -21,20 +23,62 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
-/* Taken from Mini-OS */
 
 #ifndef __UKARCH_SPINLOCK_H__
 #define __UKARCH_SPINLOCK_H__
 
+#include <uk/config.h>
+#include <stdint.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef struct {} spinlock_t;
+#ifdef CONFIG_UKPLAT_LCPU_MULTICORE
+/** spinlock struct */
+typedef struct {
+	volatile uint32_t lock;
+} __spinlock_t;
+
+/** section attribute cannot be used with pointers */
+typedef __spinlock_t *spinlock_ptr_t;
+
```

Comment:
> However, the direction into which we want the project to go is to have all relevant components SMP-safe so that spinlocks can reside in the regular .data/.bss section which is shared among all cores.

Makes sense, and it's the right approach.

> Nevertheless, we like the idea of having the option to add the attribute. So a way for OpenSynergy could be to provide an extra OpenSynergy-internal header that declares the removed DEFINE_SPINLOCK-type macro. This macro can then include the attribute to store the respective spinlock in the .slocks section as part of the variable definition.

Regarding the proposed way to override the default behavior, I can see some challenges: Overriding the macro with our own version does not cover the case where a spinlock is declared as member of another structure; you cannot use the macro there as gcc will complain that you're trying to place one member to a different section than the rest of the structure. For simple declarations of spinlocks as global variables, there is no guarantees that one would use the macro either, but that's the least of the two problems :)

Reply:
That is a fair point. Don't you have to mark those structures manually, anyway? At least your proposed solution is equivalent in this regard. How do you do at the moment?

But you are right that the `DEFINE_SPINLOCK` macro is not a good solution in that it would not cover spinlocks that we add upstream and it introduces a potential for mistakes. But at least, as long as you use it internally, your spinlocks would be covered.

Reply:
Well, the ideal would be that we somehow replaced the spinlock implementation with ours (perhaps by overriding uklock - see also https://github.com/unikraft/unikraft/discussions/261) and not patch unikraft core at all. I'm not sure if its possible though :)